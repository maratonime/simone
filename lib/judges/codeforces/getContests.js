'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getContests = getContests;

var _request = require('request');

var _request2 = _interopRequireDefault(_request);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getContests() {
  return new Promise(function (resolve, reject) {
    // making request to the codeforces api
    (0, _request2.default)('http://codeforces.com/api/contest.list', function (error, response, body) {
      if (response && response.statusCode == 200) {
        var contests = [];

        var data = JSON.parse(body).result;
        for (var index in data) {
          var info = data[index];

          // setting required fields
          var contest = {
            name: info.name,
            code: info.id,
            judge: 'codeforces',
            duration: info.durationSeconds,
            url: 'http://codeforces.com/contests/' + info.id
          };

          // setting fields that may be absent
          if ('startTimeSeconds' in info) contest.startTime = new Date(parseInt(info.startTimeSeconds) * 1000);
          if ('description' in info) contest.description = info.description;

          // setting contest state
          if (info.phase === 'BEFORE') contest.state = 'UPCOMING';else if (['CODING', 'PENDING_SYSTEM_TEST', 'SYSTEM_TEST'].includes(info.phase)) contest.state = 'RUNNING';else if (info.phase === 'FINISHED') contest.state = 'FINISHED';

          contests.push(contest);
        }
        resolve(contests);
      } else reject(error);
    });
  });
}