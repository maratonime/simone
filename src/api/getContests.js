// @flow

import type { Contest } from '../types/contest';
import type { GetContestsOptions } from '../types/api';
import judges from '../judges';

const default_options = Object.freeze({});

export async function getContests(
  options?: GetContestsOptions = default_options,
): Promise<Array<Contest>> {
  const contests: Array<Contest> = [];
  const promises: Array<Promise<Array<Contest>>> = [];
  Object.keys(judges).forEach(judge => {
    promises.push(
      judges[judge].getContests(options).catch(() => {
        console.error(judge + ' failed to getContests.');
        return [];
      }),
    );
  });
  const results = await Promise.all(promises);
  for (const contestList of results) contests.push(...contestList);
  return contests;
}
