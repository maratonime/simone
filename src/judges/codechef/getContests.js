// @flow
import { JSDOM } from 'jsdom';
import type { Contest } from '../../types';
import type { State } from '../../types';

function parse(row: HTMLTableRowElement): Contest {
  // flow has an issue concerning the cells field.
  // See https://github.com/facebook/flow/issues/6264
  const cells: HTMLCollection<HTMLTableCellElement> = (row: any).cells;

  const code: string = cells[0].textContent;
  const name: string = cells[1].children[0].textContent;
  const url: string = 'https://www.codechef.com/' + code;
  const startTime: Date = new Date(cells[2].dataset['starttime']);
  const endTime: Date = new Date(cells[3].dataset['endtime']);
  const duration: number = (endTime.getTime() - startTime.getTime()) / 1000;

  let state: State;
  const now = Date.now();
  if (startTime > now) state = 'UPCOMING';
  else if (endTime <= now) state = 'FINISHED';
  else state = 'RUNNING';

  return {
    name: name,
    code: code,
    judge: 'codechef',
    startTime: startTime,
    duration: duration,
    state: state,
    url: url,
  };
}

export async function getContests(): Promise<Array<Contest>> {
  // dom is a JSDOM, which is a third-party library that was not built with Flow
  const dom: any = await JSDOM.fromURL('https://www.codechef.com/contests');
  const document: Document = dom.window.document;

  const contestsTables: NodeList<HTMLElement> = document.querySelectorAll(
    'table.dataTable',
  );

  const contests: Array<Contest> = [];

  contestsTables.forEach(table => {
    const tab: HTMLTableElement = (table: any);
    const rows: HTMLCollection<HTMLTableRowElement> = tab.tBodies[0].rows;
    Array.from(rows).forEach(row => {
      contests.push(parse(row));
    });
  });

  return contests;
}
