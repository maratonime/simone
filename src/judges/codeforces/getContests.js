// @flow
import request from 'request-promise-native';
import type { Contest } from '../../types';
import { filterStartDate } from '../../utils';

type CFContest = {
  name: string,
  id: string,
  startTimeSeconds?: number,
  durationSeconds?: number,
  description?: string,
  phase: string,
};

export async function getContests(
  from?: Date,
  to?: Date,
): Promise<Array<Contest>> {
  const cfData: { result: Array<CFContest> } = JSON.parse(
    await request('http://codeforces.com/api/contest.list'),
  );

  return filterStartDate(
    cfData.result.map((contestData: CFContest) => {
      // setting required fields
      const contest: Contest = {
        name: contestData.name,
        code: contestData.id,
        judge: 'codeforces',
        url: 'http://codeforces.com/contests/' + contestData.id,
      };

      // setting fields that may be absent
      if (contestData.startTimeSeconds != null)
        contest.startTime = new Date(contestData.startTimeSeconds * 1000);
      if (contestData.durationSeconds != null)
        contest.duration = contestData.durationSeconds;
      if (contestData.description != null)
        contest.description = contestData.description;

      // setting contest state
      if (contestData.phase === 'BEFORE') contest.state = 'UPCOMING';
      else if (
        ['CODING', 'PENDING_SYSTEM_TEST', 'SYSTEM_TEST'].includes(
          contestData.phase,
        )
      )
        contest.state = 'RUNNING';
      else if (contestData.phase === 'FINISHED') contest.state = 'FINISHED';

      return contest;
    }),
    from,
    to,
  );
}
