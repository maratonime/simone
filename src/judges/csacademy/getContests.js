// @flow
import request from 'request-promise-native';
import type { Contest } from '../../types';

type CSContest = {|
  longName: string,
  name: string,
  id: string,
  startTime: ?string,
  endTime: ?string,
|};

export async function getContests(): Promise<Array<Contest>> {
  var options = {
    url: 'http://csacademy.com/contests/',
    headers: {
      'x-requested-with': 'XMLHttpRequest',
    },
  };
  const body = await request(options);
  const contests: Array<Contest> = [];

  const data: Object = JSON.parse(body).state.Contest;
  for (const info: CSContest of data) {
    if (
      info.longName === 'Archive' ||
      info.longName.startsWith('Virtual') || // Virtual contest for contestId=40468
      info.longName.startsWith('Contest') || // Contest for eval task 190 (Tree From Leaves)
      info.longName.match(/^\d\d:\d\d/) // 14:00, 28 September 2018 - algorithms
    )
      continue;
    // setting required fields
    const contest: Contest = {
      name: 'CSAcademy ' + info.longName,
      code: info.id,
      judge: 'csacademy',
      url: 'http://csacademy.com/contest/' + info.name,
    };

    // setting fields that may be absent
    if ('startTime' in info && info.startTime != null)
      contest.startTime = new Date(parseInt(info.startTime) * 1000);
    if (contest.startTime != null) {
      const startTime: Date = contest.startTime;
      const currentTime = Date.now();
      if ('endTime' in info && info.endTime != null) {
        const endTime: Date = new Date(parseInt(info.endTime) * 1000);
        contest.duration = Math.max(endTime - startTime, 0);
      }

      contest.state = currentTime < startTime ? 'UPCOMING' : 'RUNNING';
      if (
        contest.duration != null &&
        currentTime > new Date(startTime.getTime() + contest.duration)
      )
        contest.state = 'FINISHED';
    }

    contests.push(contest);
  }
  return contests;
}
