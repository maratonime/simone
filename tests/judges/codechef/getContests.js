import { getContests } from '../../../src/judges/codechef/getContests';
import { JSDOM } from 'jsdom';

JSDOM.fromURL = jest
  .fn()
  .mockReturnValue(JSDOM.fromFile('tests/resources/codechef.html'));

// For some reason if we mock the Date constructor jest does not work properly
const date = new Date(Date.UTC(2018, 4, 8, 10));
const old_now = Date.now;
beforeEach(() => (Date.now = () => new Date(date)));
afterEach(() => (Date.now = old_now));

test('retrieve all contests', () => {
  return expect(getContests()).resolves.toMatchSnapshot();
});

test('filtering', () => {
  return expect(
    getContests(
      new Date(date.getTime() - 1000 * 60 * 60 * 24 * 30),
      new Date(date.getTime() + 1000 * 60 * 60 * 24 * 5),
    ),
  ).resolves.toMatchSnapshot();
});
