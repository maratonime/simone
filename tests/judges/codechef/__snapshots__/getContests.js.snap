// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filtering 1`] = `
Array [
  Object {
    "code": "MAY18",
    "duration": 864000,
    "judge": "codechef",
    "name": "May Challenge 2018",
    "startTime": 2018-05-04T09:30:00.000Z,
    "state": "RUNNING",
    "url": "https://www.codechef.com/MAY18",
  },
  Object {
    "code": "LOCAPR18",
    "duration": 172800,
    "judge": "codechef",
    "name": "LoC April 2018 (Rated for Division 2)",
    "startTime": 2018-04-27T18:30:00.000Z,
    "state": "FINISHED",
    "url": "https://www.codechef.com/LOCAPR18",
  },
  Object {
    "code": "COX2018",
    "duration": 10800,
    "judge": "codechef",
    "name": "CODEX_X",
    "startTime": 2018-04-29T14:30:00.000Z,
    "state": "FINISHED",
    "url": "https://www.codechef.com/COX2018",
  },
]
`;

exports[`retrieve all contests 1`] = `
Array [
  Object {
    "code": "MAY18",
    "duration": 864000,
    "judge": "codechef",
    "name": "May Challenge 2018",
    "startTime": 2018-05-04T09:30:00.000Z,
    "state": "RUNNING",
    "url": "https://www.codechef.com/MAY18",
  },
  Object {
    "code": "INOIPRAC",
    "duration": 126230400,
    "judge": "codechef",
    "name": "INOI Practice Contest",
    "startTime": 2016-01-04T18:30:00.000Z,
    "state": "RUNNING",
    "url": "https://www.codechef.com/INOIPRAC",
  },
  Object {
    "code": "COOK94",
    "duration": 9000,
    "judge": "codechef",
    "name": "May Cook-Off 2018",
    "startTime": 2018-05-20T16:00:00.000Z,
    "state": "UPCOMING",
    "url": "https://www.codechef.com/COOK94",
  },
  Object {
    "code": "LTIME60",
    "duration": 10800,
    "judge": "codechef",
    "name": "May Lunchtime 2018",
    "startTime": 2018-05-26T14:00:00.000Z,
    "state": "UPCOMING",
    "url": "https://www.codechef.com/LTIME60",
  },
  Object {
    "code": "LOCAPR18",
    "duration": 172800,
    "judge": "codechef",
    "name": "LoC April 2018 (Rated for Division 2)",
    "startTime": 2018-04-27T18:30:00.000Z,
    "state": "FINISHED",
    "url": "https://www.codechef.com/LOCAPR18",
  },
  Object {
    "code": "COX2018",
    "duration": 10800,
    "judge": "codechef",
    "name": "CODEX_X",
    "startTime": 2018-04-29T14:30:00.000Z,
    "state": "FINISHED",
    "url": "https://www.codechef.com/COX2018",
  },
]
`;
